### ISSUE TITLE
Support for Android Auto

### ISSUE THREAD

**Issue by kaaholst** [kaaholst/android-squeezer#742]
Request from Google Play:
------------------------------------
Possibility to control Squeezer by buttons of my steering wheel. Yes, I use squeezebox in my car and it works great!

**Comment by ProteinPig** [comment]
I'm using in my car also, everything is great except I can't control through Bluetooth.

When I press "next" button, it just pause few seconds, then continue play the same song.

So I think support Bluetooth control is also great, since my car doesn't have Android Auto.

**Comment by gumil** [comment]
Is the reason for this not working because [MediaController](https://developer.android.com/reference/androidx/media2/session/MediaController) is not implemented (searching the repo did not show any files that contains MediaController)? I’d also love to have this because I would want to control music using a watch.

**Comment by explor202** [comment]
I would also like to see Squeezer be Android Auto compatible. I don't necessarily need BT support at the same time, but being able to see the app and use the touchscreen on my AA device would be sufficient for now. @kaaholst isn't that part just a change to some manifest data?

**Comment by kaaholst** [comment]
I'm afraid that making an Android Auto app is not that simple. The links below is the Google documentation on how to create an app for Android Auto. As seen in the links, a prerequisite for this that the app is a media app in the Android context. Squeezer is not that, it's a remote control for Squeezeplayers (via LMS), it does not stream media on its own. This means it's a lot of work to make this happen. Once this is in place, it is a matter of declaring support for Android Auto in the manifest, as you suggest.

[Create media apps](https://developers.google.com/cars/design/create-apps/media-apps/overview)
[Build media apps for cars](https://developer.android.com/training/cars/media)
[Add support for Android Auto to your media app](https://developer.android.com/training/cars/media/auto)

**Comment by explor202** [comment]
That's good info, thanks for sharing it. I might try to edit the APK (again) in APK Editor Studio and see what I can come up, just for my own use of course. Would you care to provide any suggestions? Or is it not that simple? I was previously able to add the different XML files, but I missed the part about declaring it a media app in the Android context.

**Comment by kaaholst** [comment]
@explor202 You are welcome to give it a try! A word of warning, though, I don't think you will have any success. I think more than editing the manifest XML is required. You also need to implement the requirements in the first two links I provided, which is a major task.

**Comment by arigit** [comment]
> I'm afraid that making an Android Auto app is not that simple. The links below is the Google documentation on how to create an app for Android Auto. As seen in the links, a prerequisite for this that the app is a media app in the Android context. Squeezer is not that, it's a remote control for Squeezeplayers (via LMS), it does not stream media on its own. This means it's a lot of work to make this happen. Once this is in place, it is a matter of declaring support for Android Auto in the manifest, as you suggest.
> 
> [Create media apps](https://developers.google.com/cars/design/create-apps/media-apps/overview) [Build media apps for cars](https://developer.android.com/training/cars/media) [Add support for Android Auto to your media app](https://developer.android.com/training/cars/media/auto)

Just to add, if I manually start playback with squeezer on the smartphone while android auto is connected, I can see in the car that the media playback is picked up, can see the artwork, track title and artist and the progress bar. The media controls (play/pause/skip) show but are unresponsive. This may be related to the lack of exposure in squeezer to media player controls to android, because my bluetooth earbuds that have play/pause and volume setting functionality are also ignored by squeezer. Other apps like spotify etc respond to those BT media player control events. I would thing that if squeezer adds support for listening to external  android media player controls, both BT Earbuds and (basic) android auto media player control will work. That would be awesome

**Comment by kaaholst** [comment]
Please check if this helps: [apk](https://drive.google.com/file/d/1xanOkuQpy36v2aw4OIm1brkCriPII6s_/view?usp=sharing). While it as not an Android Auto app, it  implement an Android Media Session, and so might be picked up. 
This version will automatically be updated when a new version (both production and beta) is released.

**Comment by explor202** [comment]
Can also try using Open Headunit as an AA emulator. I've switched to that and it supports the functions I originally requested here. https://github.com/andreknieriem/open-headunit

**Comment by arigit** [comment]
@kaaholst thanks, installed, will try this today in my car. 
For reference, as a fallback, there is also "streambridge" https://github.com/jacobpina1/StreamBridge, that is supposed to offer play/pause/skip buttons for on the AA dashboard (and also display track title and artwork) "on behalf of" any non-AA app, will test that one too.

Will report results.


For reference, two nice-to-have items for the day AA gets supported natively, important for streaming while on the move:
 
* toggle to switch on transcoding to mp3, the MIA plugin android app has this toggle. For people with hi-res lossless content this is key when on slow cellular coverage.
* caching tracks ahead of time. Apps like sympfonium support this, where you can tell the app how many tracks it should cache ahead of time (the app then downloads the tracks at the maximum speed it can and plays from it's cache), this helps with lossless streaming as often there are gaps/pauses in the stream while driving in the highway.

**Comment by arigit** [comment]
@kaaholst tested the beta version, unfortunately it did not work on AA (no sound), either on its own or via StreamBridge , However: the UI did show up with a working Play/Pause button (even with no StreamBridge)  that made Squeezer play and pause playback, but the audio did not come through to the car. Looks like the problem is on Squeezelite for Android that somehow does not send audio to the car. Same issue with LMN material app and Squeezer app, no sound via AA.

**Comment by kaaholst** [comment]
@arigit Thanks for testing.

Regarding the transcoding, caching and no sound issues, these are for the LMS player that Squeezer is controlling. As previously stated Squeezer does not currently stream music. It merely act as a remote for the players connected to LMS. 

If you have a Lyrion player, such as [squeezelite](https://f-droid.org/packages/org.lyrion.squeezelite/) or [SqueezePlayer](https://play.google.com/store/apps/details?id=de.bluegaspode.squeezeplayer) installed on the device on which Squeezer runs, it can give the impression that Squeezer plays music, but it is in fact one of these apps which does the actual streaming.
